design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,599 @@
+   directly. This will enable them to utilize the default S3 endpoint resolver in their resolver implementation.
+
+**Example: overriding the endpoint URI globally**
+
+```rust
+async fn main() {
+    let sdk_conf = aws_config::from_env().endpoint_uri("http://localhost:8123".parse().unwrap()).load().await;
+    let dynamo = aws_sdk_dynamodb::Client::new(&sdk_conf);
+    // snip ...
+}
+```
+
+**Example: overriding the endpoint resolver for a service**
+
+```rust
+/// Resolve to Localhost when an environment variable is set
+struct CustomDdbResolver;
+
+impl ResolveEndpoint<aws_sdk_dynamodb::endpoint::Params> for CustomDdbResolver {
+    fn resolve_endpoint(&self, params: &Params) -> Result<Endpoint, EndpointResolutionError> {
+        // custom resolver to redirect to DDB local if a flag is set
+        let base_endpoint = aws_sdk_dynamodb::endpoint::default_resolver().resolve_endpoint(params).expect("valid endpoint should be resolved");
+        if env::var("LOCAL") == Ok("true") {
+            // update the URI on the returned endpoint to localhost while preserving the other properties
+            base_endpoint.builder().uri("http://localhost:8888").build()
+        } else {
+            base_endpoint
+        }
+    }
+}
+
+async fn main() {
+    let conf = aws_config::load_from_env().await;
+    let ddb_conf = aws_sdk_dynamodb::config::Builder::from(&conf).endpoint_resolver(CustomDdbResolver);
+    let dynamodb = aws_sdk_dynamodb::Client::from_conf(ddb_conf);
+}
+```
+
+> **What about alternative S3 implementations? How do we say "don't put prefix bucket on this?"**
+>
+> For cases where users want to use the provided URL directly with no modification users will need to rely on service
+> specific configuration, like forcing path style addressing for S3.
+
+_Alternative Design_: [Context Aware Endpoint Trait](#context-aware-endpoint-traits)
+
+> _Optional addition_: We could add an additional `EndpointResolver` parameter to `SdkConfig` that exposed a global trait
+where `Params` is `&dyn Any` similar to [Context Aware Endpoint Trait](#context-aware-endpoint-traits). If these were
+**both** set, a runtime panic would alert users to the misconfiguration.
+
+### New Endpoint Traits
+
+The new endpoint resolution trait and `Endpoint` struct will be Smithy-native. AWS endpoint middleware will pull
+the `Endpoint`
+out of the property bag and read the properties to determine auth/signing + any other AWS metadata that may be required.
+
+An example of the `Endpoint` struct is below. This struct will be in `aws-smithy-types`, however, it should initially be
+gated with documentation warning about stability.
+
+#### The Endpoint Struct
+
+```rust
+// module: `aws_smithy_types::endpoint`
+// potential optimization to reduce / remove allocations for keys which are almost always static
+// this can also just be `String`
+type MaybeStatic<T> = Cow<'static, T>;
```

Comment:
If this is always used for `str`, why make it generic?

Reply:
no reason, can remove